File: 1342-43-2381-shifting-letters-ii/1342-43-2381-shifting-letters-ii.java
class Solution {
    public String shiftingLetters(String s, int[][] shifts) {
        int n = s.length();
        int[] v = new int[n + 1];

        for (int[] shift : shifts) {
            int l = shift[0], r = shift[1], d = shift[2];
            v[l] += d == 1 ? 1 : -1;
            if (r + 1 < n) {
                v[r + 1] += d == 1 ? -1 : 1;
            }
        }

        char[] chars = s.toCharArray();
        for (int i = 0; i < n; i++) {
            if (i > 0) {
                v[i] += v[i - 1];
            }
            chars[i] = (char) (((chars[i] - 'a' + v[i]) % 26 + 26) % 26 + 'a');
        }

        return new String(chars);
    }
}

File: 1342-43-2381-shifting-letters-ii/1342-43-2381-shifting-letters-ii.js
var shiftingLetters = function(s, shifts) {
    const n = s.length;
    const v = Array(n + 1).fill(0);

    for (const [l, r, d] of shifts) {
        v[l] += d === 1 ? 1 : -1;
        if (r + 1 < n) {
            v[r + 1] += d === 1 ? -1 : 1;
        }
    }

    const chars = s.split('');
    for (let i = 0; i < n; i++) {
        if (i > 0) {
            v[i] += v[i - 1];
        }
        chars[i] = String.fromCharCode(((chars[i].charCodeAt(0) - 'a'.charCodeAt(0) + v[i]) % 26 + 26) % 26 + 'a'.charCodeAt(0));
    }

    return chars.join('');
};
